### PR TITLE
High level skills for Habitat-llm integration

### LINKED ISSUE #15: Navigation Environment
Navigation Environment performance is worse than navigation on ASC

### LINKED ISSUE #17: Running nav/grasp/placing skills separately using LLMs
During the demo, we have need to run nav-only experiments. And we took a short-cut to integrate LLMs in nav env. We will need more clear way of doing this

### PR DESCRIPTION
**Pull request by KavitShah1998**
close #17 
close #15 

## High Level Skill Executor

There's probably a better design and this will change whenever we start adding more tools. I propose this as the first iteration in order to have an abstracted / high level skill execution.

## Usage

```
spot = SpotSkillController() 
spot.nav('kitchen_counter')
spot.pick('ball')
spot.nav('hall_table')
spot.place('hall_table')
spot.dock()
```

## Proposed development steps
- [x] Implement navigation function (`spot.nav`) 
- [x] Test navigation (`spot.nav('kitchen_counter')`) and diagnose possible mismatches between this (using `SpotNavEnv`) and the mobile manipulation env. We have already diagnosed one: #15, this is probably the only
- [x] Address diagnosed issues
- [x] Implement `.pick` (gaze)
- [x] Test and diagnose mismatches on `pick` 
- [x] Implement `.place`
- [x] Test and diagnose issues on `place`
- [ ] Incorporate and test language model for a simple skill coordination task

## Future steps

> Future tools:
> - mapping / exploration
> - open door/cabinet
> - close door/cabinet

## This PR has been moved from https://github.com/facebookresearch/spot-sim2real/pull/25

### CODE REVIEW

**Comment by KavitShah1998** [comment]
Please follow discussion on the PR : https://github.com/facebookresearch/spot-sim2real/pull/25 for context

File: spot_rl_experiments/spot_rl/envs/nav_env.py [hunk 1]
```diff
@@ -184,6 +184,22 @@ def reset(self, goal_xy, goal_heading):
...
+    #         self.goal_heading = None
+    #     else:
+    #         self.goal_xy, self.goal_heading = (waypoint[:2], waypoint[2])
+
+    #     self.goal_xy = np.array(self.goal_xy, dtype=np.float32)
+    #     self.goal_heading = self.goal_heading
+    #     observations = super().reset()
+    #     assert len(self.goal_xy) == 2
+
+    #     return observations
```
Comment (KavitShah1998) [inline spot_rl_experiments/spot_rl/envs/nav_env.py:201]: @SergioArnaud why is this commented?

Comment (KavitShah1998) [reply spot_rl_experiments/spot_rl/envs/nav_env.py]: This is not needed anymore. Removed.

File: bd_spot_wrapper/spot_wrapper/spot.py [hunk 2]
```diff
@@ -689,14 +702,60 @@ def get_base_transform_to(self, child_frame):
         ).parent_tform_child
         return kin_state.position, kin_state.rotation
 
-    def dock(self, dock_id, home_robot=False):
+    # WHY IS DEFAUULT HOME_ROBOT=FALSE???
```
Comment (jimmytyyang) [inline bd_spot_wrapper/spot_wrapper/spot.py:705]: Hi Kavit, do you want to investigate this a bit?

File: spot_rl_experiments/spot_rl/envs/base_env.py [hunk 3]
```diff
@@ -851,6 +836,7 @@ def turn_wrist(self):
         self.spot.set_arm_joint_positions(positions=arm_positions, travel_time=0.3)
         time.sleep(0.6)
 
+    # TODO: Move this to spot.py
```
Comment (jimmytyyang) [inline spot_rl_experiments/spot_rl/envs/base_env.py:839]: I think you can move this in this PR :)

File: spot_rl_experiments/spot_rl/envs/skill_manager.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+# SpotSkillManager? SpotSkillExecutor? SpotSkillController?
+class SpotSkillManager:
+    def __init__(self):
+        # We will probably receive a config as part of the constructor
+        # And eventually that specifies the set of skills that we can instantiate and execute
+        self.__init_spot()
+        self.__initiate_controllers()
+
+        # Power on the robot
+        # TODO: Check if this can be moved outside of env
```
Comment (jimmytyyang) [inline spot_rl_experiments/spot_rl/envs/skill_manager.py:38]: Hi, Kavit, do you want to clean up this?

**Review by jimmytyyang — COMMENTED** [review]
Hi Kavit,

I think the PR looks great, but we will need to clean up the TODO and some comments. Thank you, please let me know if something I can help you :)

File: spot_rl_experiments/spot_rl/envs/base_env.py (hunk 3 above)
Comment (KavitShah1998) [reply spot_rl_experiments/spot_rl/envs/base_env.py:839]: Addressed

File: spot_rl_experiments/spot_rl/envs/skill_manager.py [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+        self.lease = self.spot.get_lease(hijack=True)
+        if not self.lease:
+            print("Failed to get lease for Spot. Exiting!")
+            exit(1)
+
+        # Construct configs for nav, gaze, and place
+        self.nav_config = construct_config_for_nav()
+        self.pick_config = construct_config_for_gaze(
+            max_episode_steps=350
+        )  # TODO: Find a better episode cap
```
Comment (KavitShah1998) [inline spot_rl_experiments/spot_rl/envs/skill_manager.py:67]: @jimmytyyang  what is a good cap on episode length for pick policy?

File: spot_rl_experiments/spot_rl/envs/skill_manager.py [hunk 6]
```diff
@@ -55,14 +58,14 @@ def __init_spot(self):
 
         # Construct configs for nav, gaze, and place
         self.nav_config = construct_config_for_nav()
-        self.pick_config = construct_config_for_gaze(
-            max_episode_steps=350
-        )  # TODO: Find a better episode cap
+        self.pick_config = construct_config_for_gaze(max_episode_steps=500)
```
Comment (KavitShah1998) [inline spot_rl_experiments/spot_rl/envs/skill_manager.py:61]: @jimmytyyang 500 is the default episode length limit inside the config, no point in manually feeding it 500.

We can shrink it to 350 like Sergio did at CVPR.

File: spot_rl_experiments/spot_rl/envs/skill_manager.py [hunk 7]
```diff
@@ -95,17 +98,24 @@ def nav(self, nav_target: str = None) -> Tuple[bool, str]:
...
+            )
+            check_navigation_suc = is_pose_within_bounds(
+                result[0][-1].get("pose"),
+                _nav_target_pose_deg,
+                self.nav_config.SUCCESS_DISTANCE,
+                self.nav_config.SUCCESS_ANGLE_DIST,
+            )
+            return check_navigation_suc, "Successfully reached the target pose"
+        elif not self._should_record_trajectories:
+            return True, "Successfully reached the target pose by default"
```
Comment (KavitShah1998) [inline spot_rl_experiments/spot_rl/envs/skill_manager.py:118]: Thank you for fixing the bug Jimmy, looks good!

Did you test this on spot too?

File: spot_rl_experiments/spot_rl/envs/skill_manager.py [hunk 8]
```diff
@@ -177,8 +177,13 @@ def place(self, place_target: str = None) -> Tuple[bool, str]:
...
             return False, "Place failed to reach the target position"
 
+    def get_env(self):
+        "Get the env for the ease of the access"
+        return self.nav_controller.nav_env
+
     def dock(self):
-        # TODO: Stow back the arm
+        # Stow back the arm
+        self.get_env().reset_arm()
```
Comment (KavitShah1998) [inline spot_rl_experiments/spot_rl/envs/skill_manager.py:186]: Thanks for fixing this TODO, looks good. In future, let's add a few default configurations for arm -- READY, STOW, CARRY, etc and add a method in spot.py to bring arm to one of the defaults as per the need.

File: spot_rl_experiments/spot_rl/envs/lang_env.py [hunk 9]
```diff
@@ -428,11 +428,10 @@ def step(self, *args, **kwargs):
         try:
             main(spot, args.use_mixer, config, args.output)
         finally:
-            spot.power_off()
```
Comment (KavitShah1998) [inline spot_rl_experiments/spot_rl/envs/lang_env.py:429]: @jimmytyyang  what exactly is a Remote spot?

File: spot_rl_experiments/spot_rl/envs/skill_manager.py (hunk 7 above)
Comment (jimmytyyang) [reply spot_rl_experiments/spot_rl/envs/skill_manager.py:118]: Yes, I did! See the video on the slack channel!

File: spot_rl_experiments/spot_rl/envs/lang_env.py (hunk 9 above)
Comment (KavitShah1998) [reply spot_rl_experiments/spot_rl/envs/lang_env.py:430]: Seems like Remote Spot is the core computer connected to the spot.
Remote spot does not have method called `shutdown()`

File: spot_rl_experiments/spot_rl/envs/mobile_manipulation_env.py [hunk 10]
```diff
@@ -417,11 +417,10 @@ def step(self, *args, **kwargs):
         try:
             main(spot, args.use_mixer, config, args.output)
         finally:
-            spot.power_off()
```
Comment (KavitShah1998) [inline spot_rl_experiments/spot_rl/envs/mobile_manipulation_env.py:420]: Remote spot class does not have `shutdown()` method

**Review by jimmytyyang — APPROVED** [review]
LGTM!

File: spot_rl_experiments/spot_rl/envs/skill_manager.py (hunk 4 above)
Comment (KavitShah1998) [reply spot_rl_experiments/spot_rl/envs/skill_manager.py]: Thank you @jimmytyyang for taking care of this.

File: spot_rl_experiments/spot_rl/envs/skill_manager.py (hunk 6 above)
Comment (KavitShah1998) [reply spot_rl_experiments/spot_rl/envs/skill_manager.py]: Thanks @jimmytyyang  for addressing this.

File: spot_rl_experiments/spot_rl/envs/mobile_manipulation_env.py (hunk 10 above)
Comment (KavitShah1998) [reply spot_rl_experiments/spot_rl/envs/mobile_manipulation_env.py]: Resolved.

**Comment by KavitShah1998** [comment]
Merging the PR as it is good to merge.
Thank you @jimmytyyang  & @akshararai

**Pull request merged**